The job **`x86_64-gnu-llvm-18`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/14178218714/job/39718182690) [(plain)](https://github.com/rust-lang/rust/commit/bd29cecaf3762df0878d904267aebab94899a1d8/checks/39718182690/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#20 exporting to docker image format
#20 sending tarball 20.4s done
#20 DONE 24.8s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-18]
[CI_JOB_NAME=x86_64-gnu-llvm-18]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Listening on address 127.0.0.1:4226
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-18', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'build.print-step-timings', '--enable-verbose-tests', '--set', 'build.metrics', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: build.build          := x86_64-unknown-linux-gnu
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-18/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
[RUSTC-TIMING] cc test:false 0.583
   Compiling compiler_builtins v0.1.152
[RUSTC-TIMING] build_script_build test:false 0.292
error[E0308]: mismatched types
   --> library/core/src/num/flt2dec/strategy/grisu.rs:199:25
    |
199 |     let plus = plus.mul(&cached);
    |                     --- ^^^^^^^ expected `Fp`, found `&Fp`
    |                     |
    |                     arguments to this method are incorrect
    |
note: method defined here
   --> library/core/src/num/diy_float.rs:24:12
    |
24  |     pub fn mul(&self, other: Self) -> Self {
    |            ^^^        -----------
help: consider removing the borrow
    |
199 -     let plus = plus.mul(&cached);
199 +     let plus = plus.mul(cached);
    |

error[E0308]: mismatched types
   --> library/core/src/num/flt2dec/strategy/grisu.rs:200:27
    |
200 |     let minus = minus.mul(&cached);
    |                       --- ^^^^^^^ expected `Fp`, found `&Fp`
    |                       |
    |                       arguments to this method are incorrect
    |
note: method defined here
   --> library/core/src/num/diy_float.rs:24:12
    |
24  |     pub fn mul(&self, other: Self) -> Self {
    |            ^^^        -----------
help: consider removing the borrow
    |
200 -     let minus = minus.mul(&cached);
200 +     let minus = minus.mul(cached);
    |

error[E0308]: mismatched types
   --> library/core/src/num/flt2dec/strategy/grisu.rs:201:19
    |
201 |     let v = v.mul(&cached);
    |               --- ^^^^^^^ expected `Fp`, found `&Fp`
    |               |
    |               arguments to this method are incorrect
    |
note: method defined here
   --> library/core/src/num/diy_float.rs:24:12
    |
24  |     pub fn mul(&self, other: Self) -> Self {
    |            ^^^        -----------
help: consider removing the borrow
    |
201 -     let v = v.mul(&cached);
201 +     let v = v.mul(cached);
    |

error[E0308]: mismatched types
   --> library/core/src/num/flt2dec/strategy/grisu.rs:483:19
    |
483 |     let v = v.mul(&cached);
    |               --- ^^^^^^^ expected `Fp`, found `&Fp`
    |               |
    |               arguments to this method are incorrect
    |
note: method defined here
   --> library/core/src/num/diy_float.rs:24:12
    |
24  |     pub fn mul(&self, other: Self) -> Self {
    |            ^^^        -----------
help: consider removing the borrow
    |
483 -     let v = v.mul(&cached);
483 +     let v = v.mul(cached);
    |

For more information about this error, try `rustc --explain E0308`.
[RUSTC-TIMING] core test:false 7.517
error: could not compile `core` (lib) due to 4 previous errors
```

</details>

--- a/library/core/src/num/diy_float.rs
+++ b/library/core/src/num/diy_float.rs
@@ -21,61 +21,29 @@ pub struct Fp {
 
 impl Fp {
     /// Returns a correctly rounded product of itself and `other`.
-    pub fn mul(&self, other: &Fp) -> Fp {
-        const MASK: u64 = 0xffffffff;
-        let a = self.f >> 32;
-        let b = self.f & MASK;
-        let c = other.f >> 32;
-        let d = other.f & MASK;
-        let ac = a * c;
-        let bc = b * c;
-        let ad = a * d;
-        let bd = b * d;
-        let tmp = (bd >> 32) + (ad & MASK) + (bc & MASK) + (1 << 31) /* round */;
-        let f = ac + (ad >> 32) + (bc >> 32) + (tmp >> 32);
+    pub fn mul(self, other: Self) -> Self {
+        let (lo, hi) = self.f.widening_mul(other.f);
+        let f = hi + (lo >> 63) /* round */;
         let e = self.e + other.e + 64;
-        Fp { f, e }
+        Self { f, e }
     }
 
     /// Normalizes itself so that the resulting mantissa is at least `2^63`.
-    pub fn normalize(&self) -> Fp {
-        let mut f = self.f;
-        let mut e = self.e;
-        if f >> (64 - 32) == 0 {
-            f <<= 32;
-            e -= 32;
-        }
-        if f >> (64 - 16) == 0 {
-            f <<= 16;
-            e -= 16;
-        }
-        if f >> (64 - 8) == 0 {
-            f <<= 8;
-            e -= 8;
-        }
-        if f >> (64 - 4) == 0 {
-            f <<= 4;
-            e -= 4;
-        }
-        if f >> (64 - 2) == 0 {
-            f <<= 2;
-            e -= 2;
-        }
-        if f >> (64 - 1) == 0 {
-            f <<= 1;
-            e -= 1;
-        }
+    pub fn normalize(self) -> Self {
+        let lz = self.f.leading_zeros();
+        let f = self.f << lz;
+        let e = self.e - lz as i16;
         debug_assert!(f >= (1 << 63));
-        Fp { f, e }
+        Self { f, e }
     }
 
     /// Normalizes itself to have the shared exponent.
     /// It can only decrease the exponent (and thus increase the mantissa).
-    pub fn normalize_to(&self, e: i16) -> Fp {
+    pub fn normalize_to(self, e: i16) -> Self {
         let edelta = self.e - e;
         assert!(edelta >= 0);
         let edelta = edelta as usize;
         assert_eq!(self.f << edelta >> edelta, self.f);
-        Fp { f: self.f << edelta, e }
+        Self { f: self.f << edelta, e }
     }
 }

--- a/library/core/src/num/flt2dec/strategy/grisu.rs
+++ b/library/core/src/num/flt2dec/strategy/grisu.rs
@@ -196,9 +196,9 @@ pub fn format_shortest_opt<'a>(
     let (minusk, cached) = cached_power(ALPHA - plus.e - 64, GAMMA - plus.e - 64);
 
     // scale fps. this gives the maximal error of 1 ulp (proved from Theorem 5.1).
-    let plus = plus.mul(&cached);
-    let minus = minus.mul(&cached);
-    let v = v.mul(&cached);
+    let plus = plus.mul(cached);
+    let minus = minus.mul(cached);
+    let v = v.mul(cached);
     debug_assert_eq!(plus.e, minus.e);
     debug_assert_eq!(plus.e, v.e);
 
@@ -480,7 +480,7 @@ pub fn format_exact_opt<'a>(
     // normalize and scale `v`.
     let v = Fp { f: d.mant, e: d.exp }.normalize();
     let (minusk, cached) = cached_power(ALPHA - v.e - 64, GAMMA - v.e - 64);
-    let v = v.mul(&cached);
+    let v = v.mul(cached);
 
     // divide `v` into integral and fractional parts.
     let e = -v.e as usize;

